### PR TITLE
Receive call tree after execution

### PR DESCRIPTION
**Pull request by ureeves**
Adds the call tree to the receipt of a call, allowing the host to understand which contract spent how much gas during the total execution. This tree can be iterated through, yielding `CallTreeElem` which contain various information about the call that was made. This also means code can be written downstream that branches on what each contract spent.

Resolves: #206

### CODE REVIEW

File: piecrust/src/call_tree.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+    for child in children {
+        free_tree(child);
+    }
+}
+
+struct CallTreeNode {
+    elem: CallTreeElem,
+    children: Vec<*mut Self>,
+    prev: Option<*mut Self>,
```
Comment (moCello) [inline piecrust/src/call_tree.rs:174]: Wouldn't it make sense to say `parent` instead of `prev`?

File: piecrust/src/call_tree.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    pub spent: u64,
+    pub mem_len: usize,
+}
+
+/// The tree of contract calls.
+#[derive(Debug, Default)]
+pub struct CallTree(Option<*mut CallTreeNode>);
+
+impl CallTree {
+    /// Creates a new empty call tree, starting with at the given contract.
```
Comment (fed-franz) [inline piecrust/src/call_tree.rs:26]: with or at ?

File: piecrust/src/call_tree.rs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+impl CallTreeNode {
+    fn new(elem: CallTreeElem) -> *mut Self {
+        Box::leak(Box::new(Self {
+            elem,
+            children: Vec::new(),
+            prev: None,
+        }))
+    }
+
+    fn with_prev(elem: CallTreeElem, prev: *mut Self) -> *mut Self {
```
Comment (moCello) [inline piecrust/src/call_tree.rs:186]: same here: I find `with_parent` more descriptive

**Review by moCello — APPROVED** [review]
LGTM, just a naming nit

File: piecrust/src/call_tree.rs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    pub(crate) fn update_spent(&mut self, spent: u64) {
+        if let Some(inner) = self.0 {
+            unsafe {
+                (*inner).elem.spent = spent;
+                update_spent(inner);
+            }
+        }
+    }
+
+    /// Returns the `n`th previous element from the counting from the current
```
Comment (fed-franz) [inline piecrust/src/call_tree.rs:90]: I guess the first "from the" is a leftover

**Review by fed-franz — APPROVED** [review]
LGTM
Just a couple of typos

File: piecrust/src/call_tree.rs [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+struct CallTreeIter<'a> {
+    tree: Option<Subtree>,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a> Iterator for CallTreeIter<'a> {
+    type Item = &'a CallTreeElem;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
```
Comment (miloszm) [inline piecrust/src/call_tree.rs:208]: This is quite a long unsafe block, could it be avoided? If not, could you add a comment to it?

File: piecrust/src/call_tree.rs (hunk 1 above)
Comment (ureeves) [reply piecrust/src/call_tree.rs]: Agreed. Substituted for `parent`, including all variables and mentions in comments.

File: piecrust/src/call_tree.rs (hunk 3 above)
Comment (ureeves) [reply piecrust/src/call_tree.rs]: Ditto

File: piecrust/src/call_tree.rs (hunk 2 above)
Comment (ureeves) [reply piecrust/src/call_tree.rs]: Fixed

File: piecrust/src/call_tree.rs (hunk 4 above)
Comment (ureeves) [reply piecrust/src/call_tree.rs]: Merging the call stack with the call tree resulted in this. Fixed!

File: piecrust/src/call_tree.rs (hunk 5 above)
Comment (ureeves) [reply piecrust/src/call_tree.rs:209]: Unfortunately the traversal logic is quite heavy on the pointer dereferences, so the long unsafe block is kind of unavoidable. We also have a nodes referring to both their children *and* their parent, so pointers are unavoidable as well.

I did add a comment on why all these dereferences are safe. We ensure that the tree exists - i.e. the nodes the pointers point to exist - by leveraging a lifetime into the tree when the iterator is created.

**Review by miloszm — APPROVED** [review]

**Pull request merged**